crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -0,0 +1,286 @@
+            Some(slice) => slice,
+            None => debug_unreachable!("slice OOB: {offset}..{size}; len: {}", len),
+        }
+    }
+
+    /// Sets the `byte` at the given `index`.
+    ///
+    /// Panics when `index` is out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn set_byte(&mut self, index: usize, byte: u8) {
+        match self.current_slice_mut().get_mut(index) {
+            Some(b) => *b = byte,
+            None => debug_unreachable!("set_byte OOB: {index}; len: {}", self.len()),
+        }
+    }
+
+    /// Sets the given `value` to the memory region at the given `offset`.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn set_u256(&mut self, offset: usize, value: U256) {
+        self.set(offset, &value.to_be_bytes::<32>());
+    }
+
+    /// Set memory region at given `offset`.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn set(&mut self, offset: usize, value: &[u8]) {
+        if !value.is_empty() {
+            self.slice_mut(offset, value.len()).copy_from_slice(value);
+        }
+    }
+
+    /// Set memory from data. Our memory offset+len is expected to be correct but we
+    /// are doing bound checks on data/data_offeset/len and zeroing parts that is not copied.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn set_data(&mut self, memory_offset: usize, data_offset: usize, len: usize, data: &[u8]) {
+        if data_offset >= data.len() {
+            // nullify all memory slots
+            self.slice_mut(memory_offset, len).fill(0);
+            return;
+        }
+        let data_end = min(data_offset + len, data.len());
+        let data_len = data_end - data_offset;
+        debug_assert!(data_offset < data.len() && data_end <= data.len());
+        let data = unsafe { data.get_unchecked(data_offset..data_end) };
+        self.slice_mut(memory_offset, data_len)
+            .copy_from_slice(data);
+
+        // nullify rest of memory slots
+        // Safety: Memory is assumed to be valid. And it is commented where that assumption is made
+        self.slice_mut(memory_offset + data_len, len - data_len)
+            .fill(0);
+    }
+
+    /// Copies elements from one part of the memory to another part of itself.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn copy(&mut self, dst: usize, src: usize, len: usize) {
+        self.current_slice_mut().copy_within(src..src + len, dst);
+    }
+
+    /// Get a reference to the memory of the current context
+    #[inline(always)]
+    fn current_slice(&self) -> &[u8] {
+        // Safety: if it is a valid pointer to a slice of `self.data`
+        unsafe { slice::from_raw_parts(self.current_ptr, self.limit) }
```

Comment:
this should be current length
```suggestion
        unsafe { slice::from_raw_parts(self.current_ptr, self.current_len) }
```

Reply:
This doesn't work, or there is a misunderstanding about what this function should do. 
While also `self.limit` was not good but made it pass the tests ("wasting" memory), this should be as it is now, which is `self.data.len() - self.last_checkpoint()` (which is actually all the memory available for the current context).
With `self.current_len` tests fail for the following reason:
- when you resize, you take the range `self.current_len...new_size`, with `self.current_len` < `new_size`
- however, when you take the slice since you take it up to `self.current_len`, you get index out of bounds because it is too short